File: Kern/Tests/DuoKernTests/TrainerHintsTests.swift
import Testing
@testable import DuoKern
@testable import DuoKernTrainer

struct TrainerHintsTests {

    @Test func placeValueHintsCoverEveryDigitLength() {
        // 1 digit has no place word; 2…10 all do, for every language.
        for lang in TrainerLanguage.allCases {
            #expect(Trainer.placeValueHint(digits: 1, language: lang) == nil)
            for digits in 2...Trainer.maxLevel(kind: .numbers) {
                #expect(Trainer.placeValueHint(digits: digits, language: lang) != nil,
                        "\(lang) missing place hint for \(digits) digits")
            }
            #expect(Trainer.placeValueHint(digits: 11, language: lang) == nil)
        }
    }

    @Test func placeValueHintsAreTheExpectedWords() {
        #expect(Trainer.placeValueHint(digits: 3, language: .german) == "hundert")
        #expect(Trainer.placeValueHint(digits: 4, language: .swahili) == "elfu")
        #expect(Trainer.placeValueHint(digits: 7, language: .swahili) == "milioni")
        #expect(Trainer.placeValueHint(digits: 10, language: .german) == "Milliarde")
    }

    @Test func tensReferenceIsSwahiliOnly() {
        #expect(Trainer.tensReference(language: .german) == nil)
        #expect(Trainer.tensReference(language: .ukrainian) == nil)
        let sw = Trainer.tensReference(language: .swahili)
        #expect(sw?.count == 9)
        #expect(sw?.first == "10 kumi")
        #expect(sw?.contains("30 thelathini") == true)
        #expect(sw?.last == "90 tisini")
    }
}

File: kmp/kern/src/commonTest/kotlin/net/spross/kern/trainer/TrainerHintsTests.kt
package net.spross.kern.trainer

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class TrainerHintsTests {

    @Test
    fun placeValueHintsCoverEveryDigitLength() {
        // 1 digit has no place word; 2…10 all do, for every authored language.
        for (lang in Trainer.languages) {
            assertNull(Trainer.placeValueHint(1, lang))
            for (digits in 2..Trainer.maxLevel(TrainerKind.Numbers)) {
                assertNotNull(Trainer.placeValueHint(digits, lang), "$lang missing place hint for $digits digits")
            }
            assertNull(Trainer.placeValueHint(11, lang))
        }
    }

    @Test
    fun placeValueHintsAreTheExpectedWords() {
        assertEquals("hundert", Trainer.placeValueHint(3, "de"))
        assertEquals("elfu", Trainer.placeValueHint(4, "sw"))
        assertEquals("milioni", Trainer.placeValueHint(7, "sw"))
        assertEquals("Milliarde", Trainer.placeValueHint(10, "de"))
    }

    @Test
    fun tensReferenceIsSwahiliOnly() {
        assertNull(Trainer.tensReference("de"))
        assertNull(Trainer.tensReference("uk"))
        val sw = assertNotNull(Trainer.tensReference("sw"))
        assertEquals(9, sw.size)
        assertEquals("10 kumi", sw.first())
        assertTrue("30 thelathini" in sw)
        assertEquals("90 tisini", sw.last())
    }

    @Test
    fun unauthoredLanguagesHaveNoHintsAndNoTrainer() {
        assertNull(Trainer.placeValueHint(3, "en"))
        assertNull(Trainer.tensReference("en"))
        assertTrue(!Trainer.supports("en"))
        assertEquals(listOf("de", "sw", "uk"), Trainer.languages)
    }
}
